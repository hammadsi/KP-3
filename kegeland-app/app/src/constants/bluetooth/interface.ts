import {DeviceStackParamList} from '~routes/interface';

export type BluetoothCharacteristic = string;
export type BluetoothService = {
  uuid: string;
  characteristics: BluetoothCharacteristic[];
};

export type BluetoothProfile = {
  navScreen: keyof DeviceStackParamList;
  services: string[];
  batteryService: {
    service: BluetoothService;
    fn: (bytes: number[]) => number;
  };
};

export type PeripheralNotification = {
  value: number[];
  peripheral: string;
  characteristic: string;
  service: string;
};
