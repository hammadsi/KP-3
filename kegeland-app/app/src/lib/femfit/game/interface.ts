import Matter from 'matter-js';
import {GameEngineUpdateEventOptionType} from 'react-native-game-engine';

export type Position = {
  x: number;
  y: number;
};

export type EntityBase = {
  body: Matter.Body;
};

export type ExerciseScheme = {
  name: string;
  description: string;
  repetitions: number;
  data: [number, number][];
};

export type GameEntities = {
  [key: string]: EntityBase & {[key: string]: any};
} & {
  physics: {
    engine: Matter.Engine;
    world: Matter.World;
  };
  exercise: ExerciseScheme & {currRep: number; currStep: number};
};

export type IGameEngineSystem = (
  entities: GameEntities,
  update: GameEngineUpdateEventOptionType,
) => GameEntities;
