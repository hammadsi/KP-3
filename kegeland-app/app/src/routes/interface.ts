import {
  CompositeScreenProps,
  NavigatorScreenParams,
} from '@react-navigation/native';
import type {BottomTabScreenProps} from '@react-navigation/bottom-tabs';
import type {StackScreenProps} from '@react-navigation/stack';

/**
 * Navigation interface for auth routes
 */
export type AuthStackParamList = {
  Login: undefined;
  Register: undefined;
  'Forgot password': undefined;
};

export type DeviceStackParamList = {
  Devices: undefined;
  Femfit: undefined;
};

/**
 * Navigation interface for settings routes
 */
export type SettingsStackParamList = {
  Settings: undefined;
  'Connect device': undefined;
};

/**
 * Base navigation interface
 */
export type RootTabParamList = {
  AuthStack: NavigatorScreenParams<AuthStackParamList> | undefined;
  DeviceStack: NavigatorScreenParams<DeviceStackParamList> | undefined;
  SettingsStack: NavigatorScreenParams<SettingsStackParamList> | undefined;
};

/**
 * Interface for screens on root-level
 */
export type RootScreenProps<
  RouteName extends keyof RootTabParamList = keyof RootTabParamList,
> = BottomTabScreenProps<RootTabParamList, RouteName>;

/**
 * Interface for screens under the auth-stack
 */
export type AuthScreenProps<
  RouteName extends keyof AuthStackParamList = keyof AuthStackParamList,
> = CompositeScreenProps<
  StackScreenProps<AuthStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;

export type DeviceScreenProps<
  RouteName extends keyof DeviceStackParamList = keyof DeviceStackParamList,
> = CompositeScreenProps<
  StackScreenProps<DeviceStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;

/**
 * Interface for screens under the settings-stack
 */
export type SettingsScreenProps<
  RouteName extends keyof SettingsStackParamList = keyof SettingsStackParamList,
> = CompositeScreenProps<
  StackScreenProps<SettingsStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;
