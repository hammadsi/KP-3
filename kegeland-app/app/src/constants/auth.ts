export enum UserRole {
  PATIENT = 'patient',
  PHYSICIAN = 'physician',
}

export enum Token {
  ACCESS_TOKEN = '@access_token',
  ID_TOKEN = '@id_token',
  REFRESH_TOKEN = '@refresh_token',
}

/**
 * How often to refresh the auth tokens
 */
export const REFRESH_INTERVAL_MS = 50 * 60 * 1000; // every 50th minute
