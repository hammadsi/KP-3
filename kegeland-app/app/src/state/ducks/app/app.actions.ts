import {PayloadAction} from '@reduxjs/toolkit';

import {AppSettings, SettingsProperty} from './app.interface';

/**
 * Updates a setting in the settings state.
 * @param payload the setting to update
 * @see {@link AppSettings}
 */
export function updateSetting<T extends keyof AppSettings>(
  payload: SettingsProperty<T>,
): PayloadAction<SettingsProperty<T>> {
  return {
    type: 'updateSetting',
    payload,
  };
}
updateSetting.type = 'updateSetting';
