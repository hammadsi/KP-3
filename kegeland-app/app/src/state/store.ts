import {combineReducers, configureStore} from '@reduxjs/toolkit';
import AsyncStorage from '@react-native-async-storage/async-storage';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import reduxFlipper from 'redux-flipper';

import appReducer from './ducks/app/app.reducer';
import bluetoothReducer from './ducks/bluetooth/bluetooth.reducer';
import authReducer from './ducks/auth/auth.reducer';
import questionsReducer from './ducks/questions/questions.reducer';
import sessionReducer from './ducks/session/session.reducer';

import rootSaga, {sagaMiddleware} from './rootSaga';

/**
 * Configuration for redux-persist.
 * Redux-persist is responsible for maintaining the state
 * between sessions.
 */
const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStorage,
  whitelist: ['app', 'auth', 'bluetooth'],
};

// Initiate the root reducer
export const rootReducer = combineReducers({
  app: appReducer,
  auth: authReducer,
  bluetooth: bluetoothReducer,
  questions: questionsReducer,
  session: sessionReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

// Check if development-environment is enabled
const isDev = !['production', 'test', 'integration'].includes(
  process.env.NODE_ENV || '',
);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        })
          .concat(logger)
          .concat(reduxFlipper())
          .concat(sagaMiddleware)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(sagaMiddleware),
  devTools: isDev,
});

// Start redux-saga
sagaMiddleware.run(rootSaga);

// Create persistor for persisting state
export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
