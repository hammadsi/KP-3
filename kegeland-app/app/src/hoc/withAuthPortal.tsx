import {useFocusEffect, useNavigation} from '@react-navigation/native';
import React, {useCallback} from 'react';

import useAppDispatch from '~hooks/useAppDispatch';
import useAppSelector from '~hooks/useAppSelector';
import {setAnchorRoute} from '~state/ducks/app/app.reducer';

/**
 * Higher-order component withAuthPortal.
 * Wraps the component with an auth-guard. This guard will prevent
 * signed in users to view the wrapped component, and will be navigated
 * back instead.
 * @param Component the component to wrap
 */
const withAuthPortal =
  <P extends object>(Component: React.FC<P>): React.FC<P> =>
  (props) => {
    const navigation = useNavigation();
    const dispatch = useAppDispatch();
    const {anchorRoute} = useAppSelector((state) => state.app);
    const {isSignedIn} = useAppSelector((state) => state.auth);

    /**
     * Return either to the home screen or an anchor route if it has been set
     */
    const goBack = () => {
      if (anchorRoute) {
        const [stack, params] = anchorRoute as any;
        // Return to anchor route
        navigation.navigate(stack, params || undefined);
        // Clear the anchor route
        dispatch(setAnchorRoute(undefined));
      } else {
        // Return to home screen
        navigation.navigate('DeviceStack', {screen: 'Devices'});
      }
    };

    /**
     * Go back if the user is signed in.
     */
    useFocusEffect(
      useCallback(() => {
        if (isSignedIn) {
          goBack();
        }
      }, [isSignedIn]),
    );

    return <Component {...props} />;
  };

export default withAuthPortal;
