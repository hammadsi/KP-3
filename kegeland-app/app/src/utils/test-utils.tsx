/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
