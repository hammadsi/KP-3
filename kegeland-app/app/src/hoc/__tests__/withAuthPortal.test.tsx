import React from 'react';
import {Text} from 'react-native';
import {cloneDeep, set} from 'lodash';

import withAuthPortal from '~hoc/withAuthPortal';
import {render} from '~utils/test-utils';
import {initialState as initialAuthState} from '~state/ducks/auth/auth.reducer';
import {initialState as initialAppState} from '~state/ducks/app/app.reducer';
import {AuthState} from '~state/ducks/auth/auth.interface';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';
import {AnchorRoute, AppState} from '~state/ducks/app/app.interface';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('Test withAuthPortal-hoc', () => {
  const MockComponent: React.FC = () => {
    return <Text />;
  };

  it('should render correctly', () => {
    const Component = withAuthPortal(MockComponent);
    const component = render(<Component />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe('Navigate when user is signed in', () => {
    const authState: AuthState = {
      ...initialAuthState,
      isSignedIn: true,
    };
    const state = set(cloneDeep(initialStore), 'auth', authState);

    it('should navigate Home if no anchor route is defined', () => {
      const store = mockStore(state);
      const Component = withAuthPortal(MockComponent);
      render(<Component />, {wrapperProps: {store}});
      expect(mockedNavigate).toBeCalledWith('DeviceStack', {screen: 'Devices'});
    });

    it('should navigate to anchor route if anchor route is defined', () => {
      const anchorRoute: AnchorRoute<'SettingsStack'> = [
        'SettingsStack',
        {screen: 'Settings'},
      ];
      const appState: AppState = {...initialAppState, anchorRoute};
      const store = mockStore(set(cloneDeep(state), 'app', appState));
      const Component = withAuthPortal(MockComponent);
      render(<Component />, {wrapperProps: {store}});
      expect(mockedNavigate).toBeCalledWith(...anchorRoute);
    });
  });
});
