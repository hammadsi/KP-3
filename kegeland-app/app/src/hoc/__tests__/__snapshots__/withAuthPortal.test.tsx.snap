// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test withAuthPortal-hoc should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text />
</View>
`;
